Generalise chunk retrieval and implement per sample filtering

tomwhite opened the pull request:
Fixes #179, #180, #182, #84

This is quite a big PR, the main changes are:
* Generalise `variant_chunk_iter` in the retrieval module to support filtering by region, include/exclude expression, and samples. The chunk data dict returned contains arrays where the variants and samples dimensions have already been subsetted according to the filtering criteria.
* Add a mask to support per-sample filtering. The chunk data has an optional `call_mask` entry which is a 2D variant/sample boolean array that indicates if the corresponding variant/sample matches the filtering criteria.
* Update `vcf_writer` to use chunk data rather than reading directly from Zarr.
* Update `query` to use chunk data rather than reading directly from Zarr.
* Implement the `&&` and `||` operators from bcftools to work correctly with call masks.

The changes here will make it easy to add region/expression/sample filtering to the query (#202) and plink commands. The `variant_chunk_iter` function is also what is needed for #112.

tomwhite commented:
Thanks @jeromekelleher. No particular rush on this one, so would be good for you to spend more time looking at it. I haven't done anything for caching (that can be done later). Also, there may be some cases where filter expressions that combine 1D (variant only) and 2D (variant and sample) arrays don't work, but again I think they can be improved with more test cases later.

jeromekelleher commented:
> Also, there may be some cases where filter expressions that combine 1D (variant only) and 2D (variant and sample) arrays don't work, but again I think they can be improved with more test cases later.

Do they not work in that they raise an error, or just return the wrong value?

tomwhite commented:
Raise an error.

jeromekelleher reviewed (commented):
I've had a look through, and it looks great! I think we just need two things:

1. General cleanup of the comments/commented out tests referring to these issues we're closing (grep 180 tests/*.py picks up a few more examples)
2. A bit of additional testing coverage in bcftools_validation that checks the combination of per-sample filtering / per variant filtering with boolean operators, sample selection via -s, and region queries.

For the last, I wonder if it's worth making a synthetic VCF that has predictable values for  variant/sample combinations per sample/variant on DP and AD (say)? It's quite hard to tell from reading the examples whether these queries actually produce any output, and therefore whether we're getting meaningful test coverage. At least if we had a structured example we could craft examples that cover various corner cases and verify that we are in fact doing the same thing as bcftools.

tomwhite commented:
I've addressed these points, although we can't do comprehensive testing of region, includes, and sample filtering for queries until #202 is done - which I'd like to do separately.

> It's quite hard to tell from reading the examples whether these queries actually produce any output, and therefore whether we're getting meaningful test coverage.

Another way would be to check that there's at least one variant in the output. I've added this check to `assert_vcfs_close`.

tomwhite commented on tests/test_filter.py:
```diff
@@ -70,16 +70,90 @@ class TestFilterExpressionSample:
-            # ("FMT/GQ > 20", [0, 0, 1, 1, 1, 1, 1, 0, 0]),
-            # ("FMT/DP >= 5 && FMT/GQ > 10", [0, 0, 1, 1, 1, 0, 0, 0, 0]),
-            # ("GT > 0", [1, 1, 1, 1, 1, 0, 1, 0, 1]),
-            # ("GT > 0 & FMT/HQ >= 10", [0, 0, 1, 1, 1, 0, 0, 0, 0]),
```
Removed the `GT > 0` tests since these aren't valid bcftools filters. I get `[filter.c:3715 filter_init_] Could not parse the expression: FMT/GT > 0`
